ophyd/areadetector/filestore_mixins.py

```diff
@@ -241,25 +247,66 @@ def write_path_template(self, val):
         self._write_path_template = _ensure_trailing_slash(val)
 
     def stage(self):
-        self._point_counter = count()
+        self._datum_counter = count()
         self._locked_key_list = False
         self._datum_uids.clear()
         super().stage()
         if self.enable.get() and self._reg is None:
             raise RuntimeError('The plugin {!r} is enabled, but '
                                'the Registry (self._reg) is `None`')
+        # Subclasses will assemble resource_kwargs and call
+        # self._generate_resource(resource_kwrags)
+
+    def _generate_resource(self, resource_kwargs):
+        fn = PurePath(self._fn).relative_to(self.reg_root)
+        resource = {'spec': self.filestore_spec,
+                    'root': str(self.reg_root),
+                    'resource_path': str(fn),
+                    'resource_kwargs':resource_kwargs,
+                    'path_semantics': self.path_semantics}
+        # If a Registry is set, we need to allow it to generate the uid for us.
+        if self._reg is not None:
+            logger.debug("Inserting resource with filename %s", self._fn)
+            # register_resource has accidentally different parameter names...
+            self._resource_uid = self._reg.register_resource(
+                rpath=resource['resource_path'],
+                rkwargs=resource['resource_kwargs'],
+                root=resource['root'],
+                spec=resource['spec'],
+                path_semantics=resource['path_semantics'])
+        # If a Registry is not set, we need to generate the uid.
+        else:
+            self._resource_uid = new_uid()
+        resource['uid'] = self._resource_uid
+        self._asset_docs_cache.append(('resource', resource))
 
     def generate_datum(self, key, timestamp, datum_kwargs):
         "Generate a uid and cache it with its key for later insertion."
+        i = next(self._datum_counter)
         datum_kwargs = datum_kwargs or {}
+        datum_kwargs.update({'point_number': i})
         if self._locked_key_list:
             if key not in self._datum_uids:
                 raise RuntimeError("modifying after lock")
-        uid = self._reg.register_datum(self._resource, datum_kwargs)
-        reading = {'value': uid, 'timestamp': timestamp}
+        # This is temporarily more complicated than it will be in the
+        # future.  It needs to support old configurations that have a
+        # registry.
+        datum = {'resource': self._resource_uid,
+                 'datum_kwargs': datum_kwargs}
+        if self._reg is not None:
+            # If a Registry is set, we need to allow it to generate the
+            # datum_id for us.
+            datum_id = self._reg.register_datum(
```

Comment:
I like how by introspecting the `datum_id`, we can have a hint about how it was generated for free. That will be useful.
I would have suggested passing a new `version` kwarg in. However, this would require changing the function in `databroker`. I don't think it's worth the risk for the minimal gain it would give.

Reply:
You definitely should not depend on anything about the structure of the datum stings for _anything_.

This is the code path that is going to get eliminated eventually.